gril/grilunsol.c

```diff
@@ -416,25 +416,79 @@ int g_ril_unsol_parse_radio_state_changed(GRil *gril, const struct ril_msg *mess
 	return radio_state;
 }
 
-inline static int is_valid_strength(int signal)
+/*
+ * This function makes a similar processing to was is done by validateInput()
+ * and getLteLevel() in $AOSP/frameworks/base/telephony/java/android/telephony/
+ * SignalStrength.java. The main difference is that we linearly transform the
+ * ranges to ofono's one, while AOSP gives number of bars in a non-linear way
+ * (bins for each bar have different size). We rely on the indicator to obtain
+ * a translation to bars that makes sense for humans.
+ */
+static int get_lte_strength(int signal, int rsrp, int rssnr)
 {
-	if (signal != 99 && signal != -1)
-		return 1;
+	int s_rsrp = -1, s_rssnr = -1, s_signal = -1;
+
+	/*
+	 * The range of signal is specified to be [0, 31] by ril.h, but the code
+	 * in SignalStrength.java contradicts this.
```

Comment:
I would a reference to the spec:

/* Valid values are (0-63, 99) as defined in TS 36.331 */